_posts/2022-08-29-sbt-tutorial.md

```diff
@@ -0,0 +1,514 @@
+To support multiple versions, we can provide the settings as:
+```
+val scala212 = "2.12.16"
+val scala213 = "2.13.5"
+ThisBuild / scalaVersion := scala212
+lazy val crossproject = (project in file("crossproject"))
+  .settings(
+    crossScalaVersions := List(scala212, scala213),
+    // other settings
+  )
+```
+
+Note that, we have mentioned the default version for this project as Scala 2.12 using `ThisBuild/scalaVersion`. So when we compile, it will use scala 2.12. Since we have given the `crossScalaVersions`, we can ask sbt to compile in all supported versions. 
+To do that, we need to add the symbol `+` before sbt command:
+```
++ compile
+```
+When we publish our library, we can use `+ publish` command. This will ensure that all the supported versions of the library are published.
+
+## SBT Advanced Configurations and Options
+SBT has many configuration options and argument passing on startup. Let's look at some of the important ones.
+
+### Check SBT Version
+To check which version of sbt we are using, we can run the command:
+```
+sbt --version
+```
+
+### View Applied SBT Options
+We can view all the applied options to SBT using the debug command. 
+```
+sbt --debug
+```
+This will show a list of arguments and options applied to the sbt on startup. This is very useful in identifying if the correct parameters are applied or not.
+
+Or we can use  `sbt -v` which will also provide more information related to passed arguments, but not all debug logs.
+
+### Passing Arguments on SBT Start
+We can provide jvm arguments on sbt startup. For example, if we want to increase the heap memory of the sbt process, we can start the sbt with the parameter:
+```
+sbt -v -J-Xmx3600m
+```
+### SBT Command Alias With JVM Arguments
+We have see before on how to create SBT command aliases. We can create alias with passing jvm arguments. 
+However, we need to make sure that `fork := true` is set in build.sbt. This setting will ensure that sbt will start a forked jvm and apply the settings.
+Without a fork JVM, the jvm parameters we pass are not considered by the sbt.
+
+Let's see how to do that.
+Let's add the below line to build.sbt after adding fork:
+```
+addCommandAlias(
+  "runSpecial",
+  "; set ThisBuild/javaOptions += \"-Dport=4567\"; run"
+)
+```
+Then, we will update the Main class to read this property and print it:
+```
+val portValue = Option(System.getProperty("port"))
+println("PORT value from argument is: "+portValue)
+```
+
+Next, we need to restart sbt to get these changes to effect. Then run the command `runSpecial`. This will print the passed javaOptions.
+
+If we want to pass multiple javaOptions, we need to use a Seq as:
+addCommandAlias(
+  "runSpecial",
+  "; set ThisBuild/javaOptions ++= Seq(\"-Dport=4567\", \"-Duser=rockthejvm\"); run"
+)
+
+Now when we run `runSpecial`, the value for `user` will be None. But when we run `runSpecial2`, it will have the value for `user` as rockthejvm.
```

Comment:
Add a final chapter called Conclusion, with a single paragraph rounding up all that the article went through.